packages/backend-core/src/objectStore/objectStore.ts

```diff
@@ -503,26 +541,26 @@ export async function uploadDirectory(
   localPath: string,
   bucketPath: string
 ) {
-  bucketName = sanitizeBucket(bucketName)
-  let uploads = []
-  const files = fs.readdirSync(localPath, { withFileTypes: true })
-  for (let file of files) {
-    const path = sanitizeKey(join(bucketPath, file.name))
-    const local = join(localPath, file.name)
-    if (file.isDirectory()) {
-      uploads.push(uploadDirectory(bucketName, local, path))
-    } else {
-      uploads.push(
-        streamUpload({
+  return await tracer.trace("uploadDirectory", async span => {
+    span.addTags({ bucketName, localPath, bucketPath })
+    bucketName = sanitizeBucket(bucketName)
+    const files = await fsp.readdir(localPath, { withFileTypes: true })
+    span.addTags({ numFiles: files.length })
+    for (const file of files) {
+      const path = sanitizeKey(join(bucketPath, file.name))
+      const local = join(localPath, file.name)
+      if (file.isDirectory()) {
+        await uploadDirectory(bucketName, local, path)
+      } else {
+        await streamUpload({
           bucket: bucketName,
           filename: path,
           stream: fs.createReadStream(local),
         })
-      )
+      }
     }
-  }
-  await Promise.all(uploads)
-  return files
+    return files
+  })
```

Comment:
This function got the same treatment as `retrieveDirectory`. On top of serializing it, I have also replaced uses of `fs` with `fsp` where possible, to prevent blocking the event loop on disk IO.